src/boot/efi/boot.c

```diff
@@ -1897,6 +1897,59 @@ static VOID config_entry_add_osx(Config *config) {
         }
 }
 
+
+static VOID config_entry_add_windows(Config *config, EFI_HANDLE *device, EFI_FILE *root_dir) {
+        _cleanup_freepool_ CHAR8 *bcd = NULL;
+        const CHAR16 *title = NULL;
+        EFI_STATUS err;
+        UINTN len;
+
+        assert(config);
+        assert(device);
+        assert(root_dir);
+
+        if (!config->auto_entries)
+                return;
+
+        /* Try to find a better title. */
+        err = file_read(root_dir, L"\\EFI\\Microsoft\\Boot\\BCD", 0, 100*1024, &bcd, &len);
+        if (!EFI_ERROR(err)) {
+                static const CHAR16 *versions[] = {
+                        L"Windows 11",
+                        L"Windows 10",
+                        L"Windows 8.1",
+                        L"Windows 8",
+                        L"Windows 7",
+                        L"Windows Vista"
+                };
+
+                CHAR8 *p = bcd;
+                while (!title) {
+                        CHAR8 *q = mempmem_safe(p, len, versions[0], STRLEN(L"Windows "));
+                        if (!q)
+                                break;
+
+                        len -= q - p;
+                        p = q;
+                        if (len <= STRLEN(L"Vista"))
+                                break;
+
+                        /* We found the prefix, now try all the version strings. */
+                        for (UINTN i = 0; i < ELEMENTSOF(versions); i++) {
+                                const CHAR16 *ver = versions[i] + STRLEN("Windows ");
+                                if (CompareMem(p, ver, StrSize(ver)) == 0) {
+                                        title = versions[i];
+                                        break;
+                                }
```

Comment:
I think this would be a lot more readable with a port of the helper call memory_startswith() that we have in userspace? maybe add that? would be shorter and easier to read i think, and you could drop the explicit vista size check above